manifold/src/constructors.cpp

```diff
@@ -420,13 +420,47 @@ Manifold Manifold::Compose(const std::vector<Manifold>& manifolds) {
                       combined.halfedge_.begin() + nextEdge,
                       UpdateHalfedge({nextVert, nextEdge, nextTri}));
 
+    // Assign new IDs to triangles added in this iteration, to differentiate
+    // triangles coming from different manifolds.
+    // See the end of `boolean_result.cpp` for details.
+    VecDH<int> meshIDs;
+    VecDH<int> original;
+    for (auto &entry : impl.meshRelation_.originalID) {
+      meshIDs.push_back(entry.first | 1 << 30);
+      original.push_back(entry.second);
+    }
+    // in general, meshIDs should be small
+    thrust::sort_by_key(thrust::host, meshIDs.begin(), meshIDs.end(), original.begin());
+
+    const int numMesh = meshIDs.size();
+    const int* meshIDsPtr = meshIDs.cptrD();
+    int* originalPtr = original.ptrD();
+    int meshIDStart = combined.meshRelation_.originalID.size();
+
+    thrust::for_each(thrust::device, combined.meshRelation_.triBary.begin() + nextTri,
+                     combined.meshRelation_.triBary.begin() + nextTri +
+                         impl.meshRelation_.triBary.size(),
+                     [=] __host__ __device__(BaryRef &b) {
+                       int index = thrust::lower_bound(meshIDsPtr,
+                                                       meshIDsPtr + numMesh,
+                                                       b.meshID) -
+                                   meshIDsPtr;
+                       b.meshID = index + meshIDStart;
+                       originalPtr[index] |= 1 << 30;
+                     });
+
+    for (int i = 0; i < numMesh; ++i) {
+      if (original[i] & (1 << 30)) {
+        combined.meshRelation_.originalID[i+meshIDStart] = original[i] & ~(1 << 30);
+      }
+    }
```

Comment:
This 30-line block looks a lot like the block added to boolean-result. Can we make a function we can reuse instead?

Reply:
Yes, should be possible.